File: apps/ruby/design-pattern/05-strategy/test/strategy_test.rb
require_relative "test_helper"
require_relative "../lib/report"

class StrategyTest < Minitest::Test
  HTML_FORMATTER = lambda { |context|
    puts("<html>")
    puts("  <head>")
    puts("    <title>#{context.title}</title>")
    puts("  </head>")
    puts("  <body>")
    context.text.each { |line| puts("     <p>#{line}</p>") }
    puts("  </body>")
    puts("</html>")
  }

  def test_html_report_with_lambda
    report = Strategy::Report.new(&HTML_FORMATTER)

    assert_output(/html/) { report.output_report }
  end

  def test_plain_text_report_with_block
    report = Strategy::Report.new do |context|
      puts("***** #{context.title} *****")
      context.text.each { |line| puts(line) }
    end

    assert_output(/月次報告/) { report.output_report }
  end

  def test_formatter_can_be_swapped_at_runtime
    report = Strategy::Report.new(&HTML_FORMATTER)
    report.formatter = proc { |context| puts(context.title) }

    assert_output(/月次報告/) { report.output_report }
  end
end

File: apps/haskell/design-pattern/05-strategy/test/StrategyTest.hs
module StrategyTest (tests) where

import Data.List (isInfixOf, nub)
import Strategy
  ( Report (..)
  , formatReport
  , htmlFormatter
  , markdownFormatter
  , plainTextFormatter
  )
import Test.HUnit

tests :: Test
tests =
  TestLabel "StrategyTest" $
    TestList
      [ TestLabel "testHtmlFormatter" testHtmlFormatter
      , TestLabel "testMarkdownFormatter" testMarkdownFormatter
      , TestLabel "testSwitchStrategy" testSwitchStrategy
      ]

sampleReport :: Report
sampleReport =
  Report
    { reportTitle = "月次報告"
    , reportBody = ["順調", "問題なし"]
    }

testHtmlFormatter :: Test
testHtmlFormatter = TestCase $ do
  let result = formatReport htmlFormatter sampleReport
  assertBool "HTML 開始タグを含む" ("<html>" `isInfixOf` result)
  assertBool "HTML タイトルを含む" ("<title>月次報告</title>" `isInfixOf` result)

testMarkdownFormatter :: Test
testMarkdownFormatter = TestCase $ do
  let result = formatReport markdownFormatter sampleReport
  assertEqual
    "Markdown 形式"
    "# 月次報告\n\n- 順調\n- 問題なし\n"
    result

testSwitchStrategy :: Test
testSwitchStrategy = TestCase $ do
  let strategies = [htmlFormatter, plainTextFormatter, markdownFormatter]
      results = map (\formatter -> formatReport formatter sampleReport) strategies
  assertEqual "3 つの戦略を選べる" 3 (length results)
  assertEqual "すべて異なる結果になる" 3 (length (nub results))
